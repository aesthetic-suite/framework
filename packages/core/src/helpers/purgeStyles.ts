/* eslint-disable no-param-reassign */

export default function purgeStyles(styles: HTMLStyleElement[]) {
  styles.forEach(style => {
    style.textContent = '';

    const sheet = style.sheet as CSSStyleSheet;

    if (sheet) {
      Array.from(sheet.cssRules).forEach((rule, index) => {
        sheet.deleteRule(index);
      });
    }
  });
}
