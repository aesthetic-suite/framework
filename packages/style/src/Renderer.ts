import hash from 'string-hash';
import { isObject } from 'aesthetic-utils';
import { hyphenateProperty } from 'css-in-js-utils';
import AtomicCache from './AtomicCache';
import { ClassName, Properties, StyleParams, Property, Value } from './types';
import getDocumentStyleSheet from './getDocumentStyleSheet';
import isMediaQueryCondition from './isMediaQueryCondition';
import isSupportsCondition from './isSupportsCondition';
import isNestedSelector from './isNestedSelector';

export default class Renderer {
  protected atRuleCache = new Map();

  protected classNameCache = new AtomicCache();

  constructor() {
    // Create the elements in a strict order
    getDocumentStyleSheet('global');
    getDocumentStyleSheet('low-pri');
    getDocumentStyleSheet('high-pri');
  }

  render(properties: Properties, params: StyleParams): ClassName {
    return Object.entries(properties)
      .map(([key, value]) => {
        if (value === null || value === undefined) {
          return '';
        }

        if (isObject(value)) {
          if (isMediaQueryCondition(key) || isSupportsCondition(key)) {
            return this.render(value, { ...params, condition: key });
          } else if (isNestedSelector(key)) {
            return this.render(value, { ...params, selector: key });
          } else if (__DEV__) {
            console.warn(`Unknown property "${key}" with value "${value}".`);
          }

          return '';
        }

        return this.renderDeclaration(key as Property, value, params);
      })
      .join(' ');
  }

  renderDeclaration<K extends Property>(property: K, value: Properties[K], params: StyleParams) {
    const key = hyphenateProperty(property);
    const item = this.classNameCache.read(key, value, params);

    if (item) {
      return item.className;
    }

    const sheet = getDocumentStyleSheet(params.type);
    const rank = sheet.cssRules.length;
    const className = this.createUniqueHash(property, value, params);

    sheet.insertRule(this.createRule(className, key, value, params), rank);

    this.classNameCache.write(key, value, {
      condition: '',
      selector: '',
      type: 'low-pri',
      ...params,
      className,
      rank,
    });

    return className;
  }

  renderFontFace() {}

  renderKeyframes() {}

  protected createDeclaration(property: string, value: Value): string {
    return `${property}: ${value}`;
  }

  protected createDeclarationBlock(properties: Properties): string {
    return Object.entries(properties)
      .map(([prop, value]) => this.createDeclaration(prop, value))
      .join(';');
  }

  protected createRule(
    className: ClassName,
    property: string,
    value: Value,
    params: StyleParams,
  ): string {
    let rule = `.${className}`;

    if (params.selector) {
      rule += params.selector;
    }

    rule += `{ ${this.createDeclaration(property, value)} }`;

    if (params.condition) {
      rule = `${params.condition} { ${rule} }`;
    }

    return rule;
  }

  protected createUniqueHash(property: string, value: Value, params: StyleParams): string {
    return hash(`${params.condition || ''}${params.selector || ''}${property}${value}`);
  }
}
