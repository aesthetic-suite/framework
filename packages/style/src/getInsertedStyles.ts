import getDocumentStyleSheet from './getDocumentStyleSheet';
import { SheetType } from './types';

function extractCssText(rule: CSSStyleSheet | CSSRule): string {
  let css = rule.cssText;

  if ('cssRules' in rule) {
    const inst = (rule as unknown) as CSSGroupingRule;

    for (let i = 0; i < inst.cssRules.length; i += 1) {
      css += extractCssText(inst.cssRules[i]);
    }
  }

  return css;
}

export default function getInsertedStyles(type: SheetType): string {
  return extractCssText(getDocumentStyleSheet(type));
}
