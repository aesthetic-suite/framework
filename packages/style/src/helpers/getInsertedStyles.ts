import { arrayReduce } from '@aesthetic/utils';
import getDocumentStyleSheet from './getDocumentStyleSheet';
import { SheetType } from '../types';

export default function getInsertedStyles(type: SheetType): string {
  return arrayReduce(getDocumentStyleSheet(type).cssRules, rule => rule.cssText);
}
