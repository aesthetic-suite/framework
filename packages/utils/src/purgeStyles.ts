/* eslint-disable no-param-reassign */

import toArray from './toArray';

type RulesContainer = CSSStyleSheet | CSSGroupingRule;

// Only matches global elements
const NON_GLOBAL_PREFIX = /^(#|\.|@)/u;

function containsNestedRules(rule: unknown): rule is RulesContainer {
  return (
    (typeof CSSStyleSheet !== 'undefined' && rule instanceof CSSStyleSheet) ||
    (typeof CSSMediaRule !== 'undefined' && rule instanceof CSSMediaRule) ||
    (typeof CSSSupportsRule !== 'undefined' && rule instanceof CSSSupportsRule)
  );
}

function deleteRule(parent: RulesContainer, ruleToDelete: CSSRule) {
  Array.from(parent.cssRules).some((rule, index) => {
    if (rule === ruleToDelete && parent.cssRules[index]) {
      if (typeof parent.deleteRule === 'function') {
        parent.deleteRule(index);
      } else {
        delete parent.cssRules[index];
      }

      return true;
    }

    return false;
  });

  // CSSOM doesn't implement deleteRule(), so we delete the rule by index above.
  // Now we need to filter out the empty values or other features crash.
  if (process.env.NODE_ENV === 'test') {
    // @ts-ignore
    parent.cssRules = parent.cssRules.filter(Boolean);
  }

  // Remove this rule from the parent if there are no more child rules
  if (parent.cssRules.length === 0 && parent instanceof CSSRule) {
    if (containsNestedRules(parent.parentRule)) {
      deleteRule(parent.parentRule, parent);
    }

    if (containsNestedRules(parent.parentStyleSheet)) {
      deleteRule(parent.parentStyleSheet, parent);
    }
  }
}

function purgeRules(parent: RulesContainer, onlyGlobals: boolean = false) {
  if (!parent.cssRules) {
    return;
  }

  const rulesToDelete: CSSRule[] = [];

  // First pass to gather rules to delete. We can't delete rules within this
  // loop as indices are reset each deletion, so subsequent deletions would be
  // deleting the wrong indices. Sadly we have to do multiple passes and loops.
  Array.from(parent.cssRules).forEach(rule => {
    if (containsNestedRules(rule)) {
      purgeRules(rule);
    }

    if (onlyGlobals && NON_GLOBAL_PREFIX.test(rule.cssText)) {
      return;
    }

    rulesToDelete.push(rule);
  });

  // Second pass to actually delete the rules!
  rulesToDelete.forEach(rule => {
    deleteRule(parent, rule);
  });
}

export default function purgeStyles(
  styles: HTMLStyleElement | HTMLStyleElement[],
  onlyGlobals: boolean = false,
) {
  toArray(styles).forEach(style => {
    style.textContent = '';

    if (containsNestedRules(style.sheet)) {
      purgeRules(style.sheet, onlyGlobals);
    }
  });
}
