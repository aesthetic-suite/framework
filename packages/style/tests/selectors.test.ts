import Renderer from '../src/client/ClientRenderer';
import getInsertedStyles from '../src/helpers/getInsertedStyles';
import purgeStyles from './purgeStyles';

describe('Selectors', () => {
  let renderer: Renderer;

  beforeEach(() => {
    renderer = new Renderer();
  });

  afterEach(() => {
    purgeStyles('global');
    purgeStyles('standard');
    purgeStyles('conditions');
  });

  describe('attributes', () => {
    it('generates the correct class names with attribute selector', () => {
      const className = renderer.renderRule({
        background: '#000',
        '[disabled]': {
          backgroundColor: '#286090',
          borderColor: '#204d74',
        },
      });

      expect(className).toBe('a b c');
      expect(getInsertedStyles('standard')).toMatchSnapshot();
    });

    it('uses same class name between both APIs', () => {
      const classNameA = renderer.renderRule({
        '[disabled]': {
          backgroundColor: '#000',
        },
      });
      const classNameB = renderer.renderDeclaration('backgroundColor', '#000', {
        selector: '[disabled]',
      });

      expect(classNameA).toBe('a');
      expect(classNameA).toBe(classNameB);
      expect(getInsertedStyles('standard')).toMatchSnapshot();
    });

    it('supports complex attribute selectors', () => {
      renderer.renderDeclaration('backgroundColor', '#286090', {
        selector: '[href*="example"]',
      });

      expect(getInsertedStyles('standard')).toMatchSnapshot();
    });
  });

  describe('pseudos', () => {
    it('generates the correct class names with pseudo selector', () => {
      const className = renderer.renderRule({
        padding: '5px',
        ':hover': {
          padding: '10px',
        },
        '::before': {
          content: '"â˜…"',
          display: 'inline-block',
        },
      });

      expect(className).toBe('a b c d');
      expect(getInsertedStyles('standard')).toMatchSnapshot();
    });

    it('uses same class name between both APIs', () => {
      const classNameA = renderer.renderRule({
        ':focus': {
          backgroundColor: '#000',
        },
      });
      const classNameB = renderer.renderDeclaration('backgroundColor', '#000', {
        selector: ':focus',
      });

      expect(classNameA).toBe('a');
      expect(classNameA).toBe(classNameB);
      expect(getInsertedStyles('standard')).toMatchSnapshot();
    });

    it('supports complex attribute selectors', () => {
      renderer.renderDeclaration('color', 'white', {
        selector: ':nth-last-of-type(4n)',
      });

      expect(getInsertedStyles('standard')).toMatchSnapshot();
    });
  });

  describe('hierarchy', () => {
    it('generates the correct class names with hierarchy selector', () => {
      const className = renderer.renderRule({
        padding: '10px',
        '+ div': {
          padding: '10px',
        },
        '~ SPAN': {
          padding: '10px',
        },
        '>li': {
          padding: '10px',
        },
        '*': {
          padding: '10px',
        },
      });

      expect(className).toBe('a b c d e');
      expect(getInsertedStyles('standard')).toMatchSnapshot();
    });

    it('uses same class name between both APIs', () => {
      const classNameA = renderer.renderRule({
        '+ div': {
          backgroundColor: '#000',
        },
      });
      const classNameB = renderer.renderDeclaration('backgroundColor', '#000', {
        selector: '+ div',
      });

      expect(classNameA).toBe('a');
      expect(classNameA).toBe(classNameB);
      expect(getInsertedStyles('standard')).toMatchSnapshot();
    });

    it('supports complex attribute selectors', () => {
      renderer.renderDeclaration('color', 'white', {
        selector: ':first-of-type + li',
      });

      expect(getInsertedStyles('standard')).toMatchSnapshot();
    });
  });
});
