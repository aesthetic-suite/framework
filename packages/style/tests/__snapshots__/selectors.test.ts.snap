// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectors attributes generates the correct class names with attribute selector 1`] = `".1yedsjc {background: #000;}.1ezu8qz[disabled] {background-color: #286090;}.1m3gba5[disabled] {border-color: #204d74;}"`;

exports[`Selectors attributes supports complex attribute selectors 1`] = `".1krihrx[href*=\\"example\\"] {background-color: #286090;}"`;

exports[`Selectors attributes uses same class name between both APIs 1`] = `".4u5ry6[disabled] {background-color: #000;}"`;

exports[`Selectors hierarchy generates the correct class names with hierarchy selector 1`] = `".2jxlp9 {padding: 10px;}.azehd9+ div {padding: 10px;}.vcq0y7~ SPAN {padding: 10px;}.li1ae>li {padding: 10px;}.d8ou53* {padding: 10px;}"`;

exports[`Selectors hierarchy supports complex attribute selectors 1`] = `".q0lo5c:first-of-type + li {color: white;}"`;

exports[`Selectors hierarchy uses same class name between both APIs 1`] = `".jclqu0+ div {background-color: #000;}"`;

exports[`Selectors pseudos generates the correct class names with pseudo selector 1`] = `".l1hma1 {padding: 5px;}.9yzjcx:hover {padding: 10px;}.1dxv7sg::before {content: \\"â˜…\\";}.hmvnkb::before {display: inline-block;}"`;

exports[`Selectors pseudos supports complex attribute selectors 1`] = `".1jb0id6:nth-last-of-type(4n) {color: white;}"`;

exports[`Selectors pseudos uses same class name between both APIs 1`] = `".123nku:focus {background-color: #000;}"`;
